### PR TITLE
build a progress object on the server

### PR DESCRIPTION
**Pull request by joelhooks**
progress

### CODE REVIEW

File: apps/total-typescript/src/video/exercise-overlay.tsx [hunk 1]
```diff
@@ -310,6 +311,9 @@ const DefaultOverlay = () => {
               {lessonSlug: router.query.lesson as string},
               {
                 onSettled: (data, error, variables, context) => {
+                  utils.moduleProgress.bySlug.invalidate({
```
Comment (joelhooks) [inline apps/total-typescript/src/video/exercise-overlay.tsx:314]: this invalidates and reloads progress when the mutation succeeds 🔥

File: apps/total-typescript/src/trpc/routers/module-progress.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      return ModuleProgressSchema.parse({
+        moduleId: module._id,
+        nextLesson:
+          moduleProgressLessons.find(
+            (lesson: {lessonCompleted: boolean}) => !lesson.lessonCompleted,
+          ) || null,
+        nextSection:
+          moduleProgressSections.find(
+            (section: {sectionCompleted: boolean}) => !section.sectionCompleted,
+          ) || null,
```
Comment (joelhooks) [inline apps/total-typescript/src/trpc/routers/module-progress.ts:91]: does this need to be more complex?

File: apps/total-typescript/src/templates/workshop-template.tsx [hunk 3]
```diff
@@ -267,6 +268,12 @@ const LessonListItem = ({
     module: workshop,
   })
 
+  const moduleProgress = trpc.moduleProgress.bySlug.useQuery({
+    slug: workshop.slug.current,
+  })
+
+  console.log({moduleProgress})
```
Comment (joelhooks) [inline apps/total-typescript/src/templates/workshop-template.tsx:275]: this was just a spot to drop it in and check it out

File: apps/total-typescript/src/trpc/routers/module-progress.ts (hunk 2 above)
Comment (vojtaholik) [reply apps/total-typescript/src/trpc/routers/module-progress.ts:91]: I can't think of any reason why that would be the case

File: apps/total-typescript/src/templates/workshop-template.tsx (hunk 3 above)
Comment (vojtaholik) [reply apps/total-typescript/src/templates/workshop-template.tsx:275]: I'm gonna drop a PR testing it out!

File: apps/total-typescript/src/video/module-progress.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import {z} from 'zod'
+
+export const ModuleProgressSchema = z.object({
```
Comment (vojtaholik) [inline apps/total-typescript/src/video/module-progress.ts:3]: 🔥

**Review by vojtaholik — APPROVED** [review]

**Pull request merged**